File: 20-valid-parentheses/valid-parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        opening_brackets = {'(', '[', '{'}
        matching = {
            ')': '(',
            ']': '[',
            '}': '{'
        }
        stack = []

        for bracket in s:
            if bracket in opening_brackets:     # opening bracket
                stack.append(bracket)
            else:                               # closing bracket
                if len(stack) == 0:             # '}[]'
                    return False
                else:
                    if stack[-1] == matching[bracket]:    # matches
                        stack.pop()
                    else:
                        return False

        return len(stack) == 0

File: 20-valid-parentheses/valid-parentheses.java
class Solution {
    public boolean isValid(String s) {
        Deque<Character> stack = new ArrayDeque<>();
        Map<Character, Character> mapping = new HashMap<>();
        mapping.put(')', '(');
        mapping.put('}', '{');
        mapping.put(']', '[');

        for (char c : s.toCharArray()) {
            if (c == '(' || c == '{' || c == '[') {
                stack.push(c);
            } else {        // closing
                if (stack.isEmpty() || stack.peek() != mapping.get(c)) {
                    return false;
                }
                stack.pop();
            }
        }
        return stack.isEmpty();
    }
}